File: backend/prisma/seed.js
import bcrypt from "bcryptjs";
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

async function main() {
  const username = String(process.env.INIT_ADMIN_USERNAME || "").trim();
  const password = String(process.env.INIT_ADMIN_PASSWORD || "").trim();
  if (!username) {
    throw new Error("INIT_ADMIN_USERNAME is required");
  }

  const existing = await prisma.admin.findUnique({ where: { username } });
  if (existing) {
    // Do not rotate/reset credentials on each start.
    return;
  }

  if (!password) {
    throw new Error("INIT_ADMIN_PASSWORD is required for initial bootstrap");
  }
  const passwordHash = await bcrypt.hash(password, 10);

  await prisma.admin.create({
    data: {
      username,
      passwordHash,
      isActive: true,
      fullName: "Администратор",
      email: process.env.INIT_ADMIN_EMAIL?.trim() || null,
      inviteToken: null,
      inviteExpiresAt: null,
    },
  });
}

main()
  .catch((error) => {
    console.error("Seed failed:", error);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

File: backend/internal/seed/seed.go
package seed

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/segmentio/ksuid"
	"golang.org/x/crypto/bcrypt"

	"github.com/openvpn-control/openvpn-control-server/backend/internal/config"
)

func Run(ctx context.Context, pool *pgxpool.Pool, cfg config.Config) error {
	username := cfg.InitAdminUsername
	if username == "" {
		return fmt.Errorf("INIT_ADMIN_USERNAME is required")
	}
	var exists bool
	err := pool.QueryRow(ctx, `SELECT EXISTS(SELECT 1 FROM "Admin" WHERE username = $1)`, username).Scan(&exists)
	if err != nil {
		return err
	}
	if exists {
		return nil
	}
	password := cfg.InitAdminPassword
	if password == "" {
		return fmt.Errorf("INIT_ADMIN_PASSWORD is required for initial bootstrap")
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return err
	}
	email := cfg.InitAdminEmail
	var emailVal any
	if email != "" {
		emailVal = email
	}
	_, err = pool.Exec(ctx, `
		INSERT INTO "Admin" (id, "fullName", username, email, "passwordHash", "isActive", "createdAt", "updatedAt")
		VALUES ($1, $2, $3, $4, $5, true, NOW(), NOW())
	`, ksuid.New().String(), "Администратор", username, emailVal, string(hash))
	return err
}
